src/bridge/EspressoSGXTEEVerifier.sol

```diff
@@ -0,0 +1,171 @@
+
+/**
+ *
+ * @title  Verifies quotes from the TEE and attests on-chain
+ * @notice Contains the logic to verify a quote from the TEE and attest on-chain. It uses the V3QuoteVerifier contract
+ *         from automata to verify the quote. Along with some additional verification logic.
+ */
+contract EspressoSGXTEEVerifier is IEspressoSGXTEEVerifier, Ownable2Step {
+    using BytesUtils for bytes;
+
+    // V3QuoteVerififer contract from automata to verify the quote
+    V3QuoteVerifier public quoteVerifier;
+
+    mapping(bytes32 => bool) public registeredEnclaveHash;
+    mapping(address => bool) public registeredSigners;
+
+    constructor(bytes32 enclaveHash, address _quoteVerifier) {
+        quoteVerifier = V3QuoteVerifier(_quoteVerifier);
+        registeredEnclaveHash[enclaveHash] = true;
+    }
+
+    /*
+        @notice Verify a quote from the TEE and attest on-chain
+        The verification is considered successful if the function does not revert.
+        @param rawQuote The quote from the TEE
+        @param reportDataHash The hash of the report data
+    */
+    function verify(
+        bytes calldata rawQuote,
+        bytes32 reportDataHash
+    ) public view returns (EnclaveReport memory) {
+        // Parse the header
+        Header memory header = parseQuoteHeader(rawQuote);
+
+        // Currently only version 3 is supported
+        if (header.version != 3) {
+            revert InvalidHeaderVersion();
+        }
+
+        // Verify the quote
+        (bool success, ) = quoteVerifier.verifyQuote(header, rawQuote);
+        if (!success) {
+            revert InvalidQuote();
+        }
+
+        // Parse enclave quote
+        uint256 lastIndex = HEADER_LENGTH + ENCLAVE_REPORT_LENGTH;
+        EnclaveReport memory localReport;
+        (success, localReport) = parseEnclaveReport(rawQuote[HEADER_LENGTH:lastIndex]);
+        if (!success) {
+            revert FailedToParseEnclaveReport();
+        }
+
+        // Check that mrEnclave match
+        if (!registeredEnclaveHash[localReport.mrEnclave]) {
+            revert InvalidEnclaveHash();
+        }
+
+        //  Verify that the reportDataHash if the hash signed by the TEE
+        // We do not check the signature because `quoteVerifier.verifyQuote` already does that
+        if (reportDataHash != bytes32(localReport.reportData.substring(0, 32))) {
+            revert InvalidReportDataHash();
+        }
+
+        return localReport;
+    }
+
+    /*
+        @notice Register a new signer by verifying a quote from the TEE
+        @param attestation The attestation from the TEE
+        @param data when registering a signer, data can be passed for each TEE type
+        rest are the address
+    */
+    function registerSigner(bytes calldata attestation, bytes calldata data) external {
+        // Check that the data length is 32 bytes because verify function expects bytes32
+        if (data.length != 32) {
+            revert InvalidDataLength();
+        }
```

Comment:
Do we need to check this if we are casting the data array to bytes32? if it's shorter that would mean it gets padded by the conversion, and if the data array is longer we are truncating the key that was attested to right?
I was under the impression that ECDSA keys using secp256k1 are 33 bytes in length ([source](https://cryptobook.nakov.com/digital-signatures/ecdsa-sign-verify-messages))

Reply:
The verify function of automata expects 32 bytes. 

`I was under the impression that ECDSA keys using secp256k1 are 33 bytes in length` - yes but an Ethereum address is only 20 bytes and we are attesting that